acra-censor/handlers/handlers_util.go

```diff
@@ -283,76 +283,125 @@ func handleSelectWherePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool
+	if err != nil {
 		return false
 	}
-	// check that patterns query the same as our parsed placeholder
-	if reflect.DeepEqual(patternNode.(*sqlparser.Subquery).Select, SubqueryConfigPlaceholderReplacerParsed) {
-		return true
+	hasStar := false
+	if selectExpr, ok := patternNodes[0].(*sqlparser.Select); ok {
+		hasStar = starFound(selectExpr.SelectExprs)
 	}
-	return false
-}
 
-// matchQueryNodeWithValuePattern return true if %%VALUE%% pattern should mask value of node
-// return true for any literal values, boolean and null
-// return false on other values like subqueries
-func matchQueryNodeWithValuePattern(node sqlparser.SQLNode) bool {
-	switch node.(type) {
-	case *sqlparser.SQLVal, sqlparser.BoolVal, *sqlparser.NullVal:
-		return true
+	for i := 0; i < len(queryTopNodes); i++ {
+		patternNode := patternTopNodes[i]
+		queryNode := queryTopNodes[i]
+		switch patternNode.(type) {
+		case sqlparser.SelectExprs:
+			if hasStar {
+				// if select * in pattern then skip columns in query
+				continue
+			}
+		case *sqlparser.Where:
+			if _, ok := queryNode.(*sqlparser.Where); !ok {
+				return false
+			}
+			if !handleWhereNode(patternNode.(*sqlparser.Where), queryNode.(*sqlparser.Where)) {
+				return false
+			}
+			continue
+		}
+
+		if !reflect.DeepEqual(patternNode, queryNode) {
+			return false
+		}
 	}
-	return false
+	return true
 }
 
-// handleRangeCondition handle range queries (age BETWEEN %%value%% and 5)
-// return true if match (with or without %%value%% patterns) otherwise false
-func handleRangeCondition(patternNode, queryNode *sqlparser.RangeCond) bool {
-	if queryNode.Operator != patternNode.Operator {
+// handleLimitValuePattern evaluates LIMIT=%%VALUE%% and OFFSET=%%VALUE%% patterns
+func handleLimitValuePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
+	querySelect, ok := queryNodes[0].(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	if !reflect.DeepEqual(queryNode.Left, patternNode.Left) {
+	patternSelect, ok := patternNodes[0].(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	if !(matchValuePattern(patternNode.From, queryNode.From) || matchSubqueryPattern(patternNode.From, queryNode.From)) {
-		if !reflect.DeepEqual(patternNode.From, queryNode.From) {
-			return false
-		}
+	queryTopNodes, err := getTopNodes(queryNodes[0])
+	if err != nil {
+		return false
 	}
-	if !(matchValuePattern(patternNode.To, queryNode.To) || matchSubqueryPattern(patternNode.To, queryNode.To)) {
-		if !reflect.DeepEqual(patternNode.To, queryNode.To) {
-			return false
+	patternTopNodes, err := getTopNodes(patternNodes[0])
+	if err != nil {
+		return false
+	}
+
+	for i := 0; i < len(queryTopNodes); i++ {
+		patternNode := patternTopNodes[i]
+		queryNode := queryTopNodes[i]
+
+		switch patternNode.(type) {
+		case *sqlparser.Limit:
+			if !matchLimit(querySelect.Limit, patternSelect.Limit) {
+				return false
+			}
+		default:
+			if !reflect.DeepEqual(queryNode, patternNode) {
```

Comment:
we don't need to validate other fields of "select" strict here, only limit. we can just accept "sqlparser.Limit" type at function and work only with them and pass them from top function

Reply:
In such case it will accept any expressions except limit

Reply:
yes and we don't need to check twice other fields that can be checked in top function that call this. this function should expect only Limit and work only with this. and caller should pass correct type and work with it. it should not pass anything except limit, because other fields will be checked before (like with handleWherePatterns or something else)